File: 2026_9/2026.9.15 2 addTwoNumbers.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        current = head
        carry = 0 #进位

        while l1 is not None or l2 is not None or carry:
            # 三元判断符
            # A if 条件 else B 
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            total = val1 + val2 + carry     # 当前位求和
            carry = total // 10             # 整除计算新进位
            current_val = total % 10        # 取出当前位的结果

            # 接入新节点
            current.next = ListNode(current_val)
            current = current.next

            # 移动指针
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        return head.next

# ========== 测试代码 ========== #

# 辅助函数：将列表转为链表（头节点为第一个元素）
def list_to_linked_list(lst: List[int]) -> Optional[ListNode]:
    if not lst:
        return None
    head = ListNode(lst[0])
    current = head
    for val in lst[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

# 辅助函数：将链表转为列表（方便打印结果）
def linked_list_to_list(head: Optional[ListNode]) -> List[int]:
    result = []
    current = head
    while current:
        result.append(current.val)
        current = current.next
    return result

if __name__ == "__main__":
    solution = Solution()

     # 测试用例 1：普通情况
    # 342 + 465 = 807
    l1 = list_to_linked_list([2, 4, 3])
    l2 = list_to_linked_list([5, 6, 4])
    result = solution.addTwoNumbers(l1, l2)
    print("测试1:", linked_list_to_list(result))  # 预期输出: [7, 0, 8]
    
    # 测试用例 2：包含 0 的情况
    # 0 + 0 = 0
    l1 = list_to_linked_list([0])
    l2 = list_to_linked_list([0])
    result = solution.addTwoNumbers(l1, l2)
    print("测试2:", linked_list_to_list(result))  # 预期输出: [0]
    
    # 测试用例 3：链表长度不同且有进位
    # 9999999 + 9999 = 10009998
    l1 = list_to_linked_list([9, 9, 9, 9, 9, 9, 9])
    l2 = list_to_linked_list([9, 9, 9, 9])
    result = solution.addTwoNumbers(l1, l2)
    print("测试3:", linked_list_to_list(result))  # 预期输出: [8, 9, 9, 9, 0, 0, 0, 1]

File: 2026_9/2026.9.15 2 addTwoNumbers.c
#include <stdio.h>
#include <stdlib.h>

typedef struct ListNode{
    int val;
    struct ListNode *next;
}ListNode;

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    // 通过 malloc 创建头节点
    struct ListNode* headNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    headNode -> val = 0;
    headNode -> next = NULL;

    // 当前节点位头节点
    struct ListNode* current = headNode;

    int carry = 0;
    
    while(l1 != NULL || l2 != NULL || carry != 0){
        // 判断链表是否存在有值的节点
        int val1 = (l1 != NULL) ? (l1 -> val) : 0;
        int val2 = (l2 != NULL) ? (l2 -> val) : 0;
        
        // 两数相加
        int count = val1 + val2 + carry;    // 当前位相加总值
        carry = count / 10;             // 进位值
        int current_val = count % 10;       // 当前节点值
        /*
        eg:
        3 + 9 + 0 = 12 = count
        12 / 10 = 1 = carry
        12 % 10 = 2 = current_val
        */

        // 通过 malloc 创建新节点
        struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
        newNode -> val = current_val;
        newNode -> next = NULL;
        
        current -> next = newNode; // 将新节点链接到当前节点后面
        current = current -> next; // 将新节点作为当前节点

        if (l1 != NULL) l1 = l1 -> next;
        if (l2 != NULL) l2 = l2 -> next;
    }

    struct ListNode* result = headNode->next;
    free(headNode);
    return result;

}

// ========== 测试 ========== //

ListNode* createNode(int val) {
    ListNode* node = (ListNode*)malloc(sizeof(ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

void printList(ListNode* head) {
    while (head != NULL) {
        printf("%d", head->val);
        if (head->next != NULL) printf(" -> ");
        head = head->next;
    }
    printf("\n");
}

void freeList(ListNode* head) {
    while (head != NULL) {
        ListNode* temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    // 测试用例: 342 + 465 = 807
    // 链表表示为逆序: (2 -> 4 -> 3) + (5 -> 6 -> 4) = (7 -> 0 -> 8)
    
    // 创建链表 1: 342
    ListNode* l1 = createNode(2);
    l1->next = createNode(4);
    l1->next->next = createNode(3);

    // 创建链表 2: 465
    ListNode* l2 = createNode(5);
    l2->next = createNode(6);
    l2->next->next = createNode(4);

    printf("链表 1: ");
    printList(l1);
    printf("链表 2: ");
    printList(l2);

    // 调用两数相加函数
    ListNode* result = addTwoNumbers(l1, l2);

    printf("相加结果: ");
    printList(result); // 期望输出: 7 -> 0 -> 8

    // 释放所有分配的内存
    freeList(l1);
    freeList(l2);
    freeList(result);

    return 0;
}